// @atlaskit/pagination/dist/cjs/pagination.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _react = require("react");

var _core = require("@emotion/core");

var _analyticsNext = require("@atlaskit/analytics-next");

var _noop = _interopRequireDefault(require("@atlaskit/ds-lib/noop"));

var _useControlled3 = _interopRequireDefault(require("@atlaskit/ds-lib/use-controlled"));

var _chevronLeftLarge = _interopRequireDefault(require("@atlaskit/icon/glyph/chevron-left-large"));

var _chevronRightLarge = _interopRequireDefault(require("@atlaskit/icon/glyph/chevron-right-large"));

var _navigator = _interopRequireDefault(require("./internal/components/navigator"));

var _page = _interopRequireDefault(require("./internal/components/page"));

var _renderEllipsis = _interopRequireDefault(require("./internal/components/render-ellipsis"));

var _constants = require("./internal/constants");

var _styles = require("./internal/styles");

var _collapseRange = _interopRequireDefault(require("./internal/utils/collapse-range"));

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) { symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); } keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

var analyticsAttributes = {
  componentName: 'pagination',
  packageName: "@atlaskit/pagination",
  packageVersion: "14.0.2"
};

function InnerPagination(_ref, ref) {
  var _ref$components = _ref.components,
      components = _ref$components === void 0 ? _constants.emptyObject : _ref$components,
      _ref$defaultSelectedI = _ref.defaultSelectedIndex,
      defaultSelectedIndex = _ref$defaultSelectedI === void 0 ? 0 : _ref$defaultSelectedI,
      selectedIndex = _ref.selectedIndex,
      _ref$label = _ref.label,
      label = _ref$label === void 0 ? 'pagination' : _ref$label,
      _ref$previousLabel = _ref.previousLabel,
      previousLabel = _ref$previousLabel === void 0 ? 'previous' : _ref$previousLabel,
      _ref$nextLabel = _ref.nextLabel,
      nextLabel = _ref$nextLabel === void 0 ? 'next' : _ref$nextLabel,
      _ref$style = _ref.style,
      style = _ref$style === void 0 ? _constants.emptyObject : _ref$style,
      _ref$max = _ref.max,
      max = _ref$max === void 0 ? 7 : _ref$max,
      _ref$onChange = _ref.onChange,
      onChange = _ref$onChange === void 0 ? _noop.default : _ref$onChange,
      pages = _ref.pages,
      getPageLabel = _ref.getPageLabel,
      _ref$renderEllipsis = _ref.renderEllipsis,
      renderEllipsis = _ref$renderEllipsis === void 0 ? _renderEllipsis.default : _ref$renderEllipsis,
      analyticsContext = _ref.analyticsContext,
      testId = _ref.testId;

  var _useControlled = (0, _useControlled3.default)(selectedIndex, function () {
    return defaultSelectedIndex || 0;
  }),
      _useControlled2 = (0, _slicedToArray2.default)(_useControlled, 2),
      selectedIndexValue = _useControlled2[0],
      setSelectedIndexValue = _useControlled2[1];

  var onChangeWithAnalytics = (0, _analyticsNext.usePlatformLeafEventHandler)(_objectSpread({
    fn: function fn(value, analyticsEvent) {
      var event = value.event,
          selectedPageIndex = value.selectedPageIndex;

      if (selectedIndex === undefined) {
        setSelectedIndexValue(selectedPageIndex);
      }

      if (onChange) {
        onChange(event, pages[selectedPageIndex], analyticsEvent);
      }
    },
    action: 'changed',
    actionSubject: 'pageNumber',
    analyticsData: analyticsContext
  }, analyticsAttributes));

  var transform = function transform(page, currPageIndex) {
    var selectedPage = pages[selectedIndexValue];
    return (0, _core.jsx)(_page.default, {
      key: "page-".concat(getPageLabel ? getPageLabel(page, currPageIndex) : currPageIndex),
      component: components.Page,
      onClick: function onClick(event) {
        return onChangeWithAnalytics({
          event: event,
          selectedPageIndex: currPageIndex
        });
      },
      isSelected: page === selectedPage,
      page: page
    }, getPageLabel ? getPageLabel(page, currPageIndex) : page);
  };

  return (0, _core.jsx)("nav", {
    "data-testid": testId,
    css: _objectSpread(_objectSpread({}, _styles.navStyle), style),
    ref: ref,
    "aria-label": label
  }, (0, _core.jsx)(_navigator.default, {
    key: "left-navigator",
    component: components.Previous,
    onClick: function onClick(event) {
      return onChangeWithAnalytics({
        event: event,
        selectedPageIndex: selectedIndexValue - 1
      });
    },
    isDisabled: selectedIndexValue === 0,
    iconBefore: (0, _core.jsx)(_chevronLeftLarge.default, {
      label: ""
    }),
    "aria-label": previousLabel,
    pages: pages
  }), (0, _collapseRange.default)(pages, selectedIndexValue, {
    max: max,
    ellipsis: renderEllipsis,
    transform: transform
  }), (0, _core.jsx)(_navigator.default, {
    key: "right-navigator",
    component: components.Next,
    onClick: function onClick(event) {
      return onChangeWithAnalytics({
        event: event,
        selectedPageIndex: selectedIndexValue + 1
      });
    },
    isDisabled: selectedIndexValue === pages.length - 1,
    iconBefore: (0, _core.jsx)(_chevronRightLarge.default, {
      label: ""
    }),
    "aria-label": nextLabel,
    pages: pages
  }));
}

var Pagination = /*#__PURE__*/(0, _react.forwardRef)(InnerPagination);

var _default = /*#__PURE__*/(0, _react.memo)(Pagination);

exports.default = _default;
