// @atlaskit/avatar/dist/cjs/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ICON_OFFSET = exports.ICON_SIZES = exports.ACTIVE_SCALE_FACTOR = exports.BORDER_WIDTH = exports.AVATAR_RADIUS = exports.AVATAR_SIZES = void 0;

var _constants = require("@atlaskit/theme/constants");

// eslint-disable-next-line @repo/internal/fs/filename-pattern-match
var AVATAR_SIZES = {
  xsmall: (0, _constants.gridSize)() * 2,
  small: (0, _constants.gridSize)() * 3,
  medium: (0, _constants.gridSize)() * 4,
  large: (0, _constants.gridSize)() * 5,
  xlarge: (0, _constants.gridSize)() * 12,
  xxlarge: (0, _constants.gridSize)() * 16
}; // border radius only applies to "square" avatars

exports.AVATAR_SIZES = AVATAR_SIZES;
var AVATAR_RADIUS = {
  xsmall: 2,
  small: 2,
  medium: 3,
  large: 3,
  xlarge: 6,
  xxlarge: 12
};
exports.AVATAR_RADIUS = AVATAR_RADIUS;
var BORDER_WIDTH = 2;
exports.BORDER_WIDTH = BORDER_WIDTH;
var ACTIVE_SCALE_FACTOR = 0.9;
exports.ACTIVE_SCALE_FACTOR = ACTIVE_SCALE_FACTOR;
var ICON_SIZES = {
  small: 12,
  medium: 14,
  large: 15,
  xlarge: 18
};
exports.ICON_SIZES = ICON_SIZES;
var ICON_OFFSET = {
  small: 0,
  medium: 0,
  large: 1,
  xlarge: 7
};
exports.ICON_OFFSET = ICON_OFFSET;
