// @atlaskit/avatar/dist/cjs/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "default", {
  enumerable: true,
  get: function get() {
    return _Avatar.default;
  }
});
Object.defineProperty(exports, "AvatarItem", {
  enumerable: true,
  get: function get() {
    return _AvatarItem.default;
  }
});
Object.defineProperty(exports, "Presence", {
  enumerable: true,
  get: function get() {
    return _Presence.default;
  }
});
Object.defineProperty(exports, "Status", {
  enumerable: true,
  get: function get() {
    return _Status.default;
  }
});
Object.defineProperty(exports, "Skeleton", {
  enumerable: true,
  get: function get() {
    return _Skeleton.default;
  }
});
Object.defineProperty(exports, "AVATAR_SIZES", {
  enumerable: true,
  get: function get() {
    return _constants.AVATAR_SIZES;
  }
});
Object.defineProperty(exports, "BORDER_WIDTH", {
  enumerable: true,
  get: function get() {
    return _constants.BORDER_WIDTH;
  }
});
Object.defineProperty(exports, "AVATAR_RADIUS", {
  enumerable: true,
  get: function get() {
    return _constants.AVATAR_RADIUS;
  }
});
Object.defineProperty(exports, "ACTIVE_SCALE_FACTOR", {
  enumerable: true,
  get: function get() {
    return _constants.ACTIVE_SCALE_FACTOR;
  }
});

var _Avatar = _interopRequireDefault(require("./Avatar"));

var _AvatarItem = _interopRequireDefault(require("./AvatarItem"));

var _Presence = _interopRequireDefault(require("./Presence"));

var _Status = _interopRequireDefault(require("./Status"));

var _Skeleton = _interopRequireDefault(require("./Skeleton"));

var _constants = require("./constants");
