// @atlaskit/item/dist/cjs/components/ItemGroup.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireWildcard(require("react"));

var _reactAddonsTextContent = _interopRequireDefault(require("react-addons-text-content"));

var _ItemGroup = require("../styled/ItemGroup");

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

var ItemGroup = /*#__PURE__*/function (_Component) {
  (0, _inherits2.default)(ItemGroup, _Component);

  var _super = _createSuper(ItemGroup);

  function ItemGroup() {
    (0, _classCallCheck2.default)(this, ItemGroup);
    return _super.apply(this, arguments);
  }

  (0, _createClass2.default)(ItemGroup, [{
    key: "render",
    value: function render() {
      var _this = this;

      var _this$props = this.props,
          children = _this$props.children,
          elemAfter = _this$props.elemAfter,
          isCompact = _this$props.isCompact,
          title = _this$props.title,
          label = _this$props.label,
          innerRef = _this$props.innerRef,
          role = _this$props.role;

      var ariaLabel = function () {
        if (label) {
          return (0, _reactAddonsTextContent.default)(label);
        }

        if (title) {
          return (0, _reactAddonsTextContent.default)(title);
        }

        return '';
      }();

      return /*#__PURE__*/_react.default.createElement("div", {
        "aria-label": ariaLabel,
        role: role,
        ref: innerRef
      }, title ? /*#__PURE__*/_react.default.createElement(_ItemGroup.GroupTitle, {
        "aria-hidden": "true",
        isCompact: isCompact
      }, /*#__PURE__*/_react.default.createElement(_ItemGroup.GroupTitleText, null, title), elemAfter ? /*#__PURE__*/_react.default.createElement(_ItemGroup.GroupTitleAfter, {
        innerRef: function innerRef(r) {
          _this.headingAfterElement = r;
        }
      }, elemAfter) : null) : null, children);
    }
  }]);
  return ItemGroup;
}(_react.Component);

exports.default = ItemGroup;
(0, _defineProperty2.default)(ItemGroup, "defaultProps", {
  role: 'group'
});
